src/db/collections/outputs/mod.rs

```diff
@@ -365,3 +368,419 @@ impl MongoDb {
+                    None,
+                )
+                .await?
+                .try_next()
+                .await?
+                .map(bson::from_document::<StorageDepositAnalytics>)
+                .transpose()?
+                .unwrap_or_default();
+
+                Ok(Some(StorageDepositAnalyticsResult {
+                    output_count: res.output_count,
+                    storage_deposit_return_count: res.storage_deposit_return_count,
+                    storage_deposit_return_total_value: res.storage_deposit_return_total_value,
+                    total_key_bytes: res.total_key_bytes,
+                    total_data_bytes: res.total_data_bytes,
+                    total_byte_cost: res.total_byte_cost,
+                    ledger_index,
+                    rent_structure,
+                }))
+            } else {
+                Ok(None)
+            }
+        } else {
+            Ok(None)
+        }
+    }
+}
+
+/// Address analytics result.
+
+#[derive(Copy, Clone, Debug, Serialize, Deserialize)]
+pub struct AddressAnalyticsResult {
+    /// The number of addresses used in the time period.
+    pub total_active_addresses: u64,
+    /// The number of addresses that received tokens in the time period.
+    pub receiving_addresses: u64,
+    /// The number of addresses that sent tokens in the time period.
+    pub sending_addresses: u64,
+}
+
+impl MongoDb {
+    /// Create aggregate statistics of all addresses.
+    pub async fn get_address_analytics(
+        &self,
+        start_timestamp: Option<MilestoneTimestamp>,
+        end_timestamp: Option<MilestoneTimestamp>,
+    ) -> Result<Option<AddressAnalyticsResult>, Error> {
+        Ok(self
+            .db
+            .collection::<AddressAnalyticsResult>(OutputDocument::COLLECTION)
+            .aggregate(
+                vec![
+                    doc! { "$match": {
+                        "details.address": { "$exists": true }
+                    } },
+                    doc! { "$facet": {
+                        "total": [
+                            { "$match": {
+                                "$or": [
+                                    { "$nor": [
+                                        { "metadata.booked.milestone_timestamp": { "$lt": start_timestamp } },
+                                        { "metadata.booked.milestone_timestamp": { "$gte": end_timestamp } },
+                                    ] },
+                                    { "$nor": [
+                                        { "metadata.spent_metadata.spent.milestone_timestamp": { "$lt": start_timestamp } },
+                                        { "metadata.spent_metadata.spent.milestone_timestamp": { "$gte": end_timestamp } },
+                                    ] },
+                                ],
+                            } },
+                            { "$group" : { "_id": "$details.address" }},
+                            { "$count": "addresses" },
```

Comment:
Aren't we counting addresses multiple times here? I would expect some form of `distinct` or `addToSet` call here.

Reply:
That's what the `$group` does?